### PR TITLE
#162948840 implement social authentication with Google, Facebook, and Twitter

### PR DESCRIPTION
**Pull request by Georgeygigz**
**What does this PR do?**
Implement social login with Google, Facebook, and Twitter

Description of Task to be completed?
- Implement social login functionalities for oauth1 and oauth2
- write tests for social auth functionality

**How should this be manually tested?**
- 'git clone https://github.com/andela/ah-django-unchained' 

- `cd ah-django-unchained`

- `git checkout ft-social-authentication-162948840`

- create a virtual env `python3 -m venv virtual_env_name`

- Add the following congigurations to .env file
```
export FACEBOOK_APP_ID="facebook_id"
export FACEBOOK_APP_SECRET="facebook_secret"
export GOOGLE_OAUTH2_KEY="google_key"
export OAUTH2_SECRET="google_secret"
export TWITTER_KEY="twitter_key" 
export TWITTER_SECRET="twitter_secret"
```
- `source .env`

- `pip install -r requirements.txt`

**Start in postman using the following url**
- http://127.0.0.1:8000/api/login/oauth/

**To test different login funcionalities:**

- Facebook
```
{
    "provider":"facebook",
    "access_token":"facebook_access_token"
}
```
- Google:
```
{
    "provider":"google-oauth2",
    "access_token":"google_access_token"
}
```
- Twitter
```
{
    "provider": "twitter",
    "access_token": "access_token",
    "access_token_secret": "access_token_secret"

}
```
**Screenshots**
image

image

image

### CODE REVIEW

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -21,32 +24,27 @@ def authenticate(self, request):
...
+    def generate_token(email, username):
+        """Generate token."""
+        date = datetime.now() + timedelta(days=20)
+        payload = {
+            'email': email,
+            'username': username,
+            'exp': int(date.strftime('%s'))
+        }
+        token = jwt.encode(payload, settings.SECRET_KEY, algorithm='HS256')
+        return token.decode()
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/backends.py]: Are you supposed to the token here?

Comment (Georgeygigz) [reply authors/apps/authentication/backends.py]: Ignatius, Did you meant decode

File: .gitignore [hunk 2]
```diff
@@ -6,6 +6,8 @@ __pycache__/
 # C extensions
 *.so
 *.vscode
+**/migrations/*
```
Comment (ignatiusukwuoma) [inline .gitignore]: It's fine to commit migrations to version control, you don't need these

**Comment by kwameasiago** [comment]
Why is your Travis failing ?

File: authors/apps/authentication/tests/test_social_auth.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
+import json
+from django.urls import reverse
+from rest_framework.views import status
+from rest_framework.test import APITestCase, APIClient
+
+
+class SocialAuthTest(APITestCase):
+    """Test socil authentixation functionality"""
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/tests/test_social_auth.py]: typos

File: .gitignore (hunk 2 above)
Comment (Georgeygigz) [reply .gitignore]: Okay have worked on that

File: authors/apps/authentication/tests/test_social_auth.py (hunk 3 above)
Comment (Georgeygigz) [reply authors/apps/authentication/tests/test_social_auth.py]: Okay working on this

**Comment by Georgeygigz** [comment]
> Why is your Travis failing?
There is a problem with my environment variables.

File: authors/apps/authentication/views.py [hunk 4]
```diff
@@ -1,24 +1,33 @@
...
 from rest_framework.views import APIView
-from rest_framework.renderers import BrowsableAPIRenderer
-from .renderers import UserJSONRenderer
+from social_django.utils import load_strategy, load_backend
+from social_core.exceptions import MissingBackend, AuthAlreadyAssociated
+from social_core.backends.oauth import BaseOAuth1, BaseOAuth2
+
+from authors.apps.core.permissions import IsAuthorOrReadOnly
+from authors.apps.authentication.backends import JWTAuthentication
 from . import models
```
Comment (MaggieChege) [inline authors/apps/authentication/views.py:18]: This has been imported in line 21. serializers have also been imported twice.

File: authors/settings.py [hunk 5]
```diff
@@ -171,7 +209,7 @@
 }
 
 
-SECRET_KEY = os.getenv('SECRET_KEY')
+SECRET_KEY = os.getenv("SECRET_KEY")
```
Comment (kwameasiago) [inline authors/settings.py]: Hey do not mix ' and " in your code

Comment (Georgeygigz) [reply authors/settings.py]: Noted

File: authors/apps/authentication/views.py (hunk 4 above)
Comment (Georgeygigz) [reply authors/apps/authentication/views.py:18]: Okay fixing it

File: .travis.yml [hunk 6]
```diff
@@ -11,9 +11,15 @@ install:
 services: 
   - postgresql
 
+addons:
+  postgresql: '9.4'
+  apt:
+    packages:
+      - postgresql-server-dev-9.4
+
```
Comment (ignatiusukwuoma) [inline .travis.yml:16]: What does this do?

File: authors/apps/authentication/migrations/0001_initial.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 2.1.4 on 2019-01-08 14:41
+# Generated by Django 2.1.4 on 2019-01-18 07:38
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/migrations/0001_initial.py:1]: We shouldn't update the migration files

Comment (Georgeygigz) [reply authors/apps/authentication/migrations/0001_initial.py:1]: Okay noted

File: .travis.yml (hunk 6 above)
Comment (Georgeygigz) [reply .travis.yml:16]: It updates the postgress database to be used in travis

File: authors/apps/authentication/serializers.py [hunk 8]
```diff
@@ -195,3 +195,9 @@ def update(self, instance, validated_data):
         instance.save()
 
         return instance
+
+
+class SocialAuthSerializer(serializers.Serializer):
+    """Social auth serializers."""
+    provider = serializers.CharField(max_length=255, required=True)
+    access_token = serializers.CharField(max_length=1024, required=True, trim_whitespace=True)
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/serializers.py]: Leave a blank line at the end of all your files

Comment (Georgeygigz) [reply authors/apps/authentication/serializers.py]: Okay will do that

**Comment by ignatiusukwuoma** [comment]
Have you been able to get the user's email? Have you tried making requesting it from the provider using the access-token?

File: authors/apps/authentication/views.py [hunk 9]
```diff
@@ -151,3 +159,71 @@ def put(self, request, token, **kwargs):
...
+            return Response({
+                "errors": "User is authenticated"},
+                status=status.HTTP_400_BAD_REQUEST)
+        except BaseException:
+            return Response({
+                "errors": "The token has expired"},
+                 status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+        if user:
+            user.is_active = True
+            username = user.username
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/views.py:256]: What happens if username or email already exists. You have to handle it.

**Comment by ignatiusukwuoma** [comment]
I would like to understand what you have in mind on the social login flow. Explain to me, what will happen after user clicks on a button `Login with Facebook` on the frontend

File: authors/apps/authentication/views.py (hunk 9 above)
Comment (Georgeygigz) [reply authors/apps/authentication/views.py:256]: I have handled this, already

**Review by MaggieChege — APPROVED** [review]
Look good to me

File: .travis.yml [hunk 10]
```diff
@@ -12,11 +12,14 @@ services:
   - postgresql
 
 addons:
-  postgresql: "9.6"
+  postgresql: '9.4'
```
Comment (ignatiusukwuoma) [inline .travis.yml]: I think someone updated it to 9.6. Do you want to downgrade it?

File: .gitignore [hunk 11]
```diff
@@ -26,6 +26,7 @@ var/
 *.egg
 .idea/
 
+
```
Comment (ignatiusukwuoma) [inline .gitignore:29]: Remove this blank line from this file

File: authors/apps/authentication/migrations/0001_initial.py (hunk 7 above)
Comment (ignatiusukwuoma) [reply authors/apps/authentication/migrations/0001_initial.py:1]: Can you just remove the change made to this file?

File: authors/apps/authentication/views.py [hunk 12]
```diff
@@ -1,25 +1,33 @@
...
 from rest_framework.permissions import AllowAny, IsAuthenticated
 from rest_framework.response import Response
 from rest_framework.views import APIView
-from rest_framework.renderers import BrowsableAPIRenderer
-from .renderers import UserJSONRenderer
+
+
 from . import models
+from .renderers import UserJSONRenderer
+from . import serializers
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/views.py]: Unused import

File: authors/apps/authentication/views.py [hunk 13]
```diff
@@ -1,25 +1,33 @@
...
+from social_core.backends.oauth import BaseOAuth1, BaseOAuth2
+
 from django.conf import settings
 from django.core.mail import send_mail
 from django.template.loader import render_to_string
+from django.db import IntegrityError
 from rest_framework import status, generics
 from rest_framework.permissions import AllowAny, IsAuthenticated
 from rest_framework.response import Response
 from rest_framework.views import APIView
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/views.py:15]: unused import

File: .gitignore (hunk 11 above)
Comment (Georgeygigz) [reply .gitignore:29]: Okay working on it

File: .travis.yml (hunk 10 above)
Comment (Georgeygigz) [reply .travis.yml]: No, let me use the 9.6

File: authors/apps/authentication/views.py (hunk 12 above)
Comment (Georgeygigz) [reply authors/apps/authentication/views.py]: Okay working on it

File: authors/apps/authentication/views.py (hunk 13 above)
Comment (Georgeygigz) [reply authors/apps/authentication/views.py:15]: I Am using the import line 246

File: authors/apps/authentication/tests/test_social_auth.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
+import json
+import os
+from django.urls import reverse
+from rest_framework.views import status
+from rest_framework.test import APITestCase, APIClient
+
+
+class SocialAuthTest(APITestCase):
+    """Test social authentication funcionality functionality"""
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/tests/test_social_auth.py]: Remove unnecessary word

File: authors/apps/authentication/tests/test_social_auth.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        """Test login with google"""
+        response = self.client.post(
+            self.social_oauth_url, self.google_provider, format='json')
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+    
+    def test_login_with_facebook(self):
+        """Test login with facebook."""
+        response = self.client.post(
+            self.social_oauth_url, self.facebook_provider, format='json')
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/tests/test_social_auth.py:60]: How about checking that the content of response.data is what was expected for the tests above?

Comment (Georgeygigz) [reply authors/apps/authentication/tests/test_social_auth.py:60]: Okay will work on that

File: authors/apps/authentication/tests/test_social_auth.py (hunk 14 above)
Comment (Georgeygigz) [reply authors/apps/authentication/tests/test_social_auth.py]: okay

File: authors/apps/authentication/tests/test_social_auth.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+                                    format='json')
+        self.assertEqual(json.loads(response.content),
+                         {"errors": {"access_token":
+                          ["This field is required."]}})
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_rejects_login_missing_provider_name(self):
+        """Test missing provider."""
+        response = self.client.post(self.social_oauth_url,
+                                    data={"access_token":
```
Comment (ignatiusukwuoma) [inline authors/apps/authentication/tests/test_social_auth.py]: try to have key-value pairs on the same line except line is too long

Comment (Georgeygigz) [reply authors/apps/authentication/tests/test_social_auth.py]: Okay

**Pull request merged**